src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -95,19 +120,59 @@ public Http2Connection(Http2ConnectionContext context)
 
         public void OnInputOrOutputCompleted()
         {
-            _stopping = true;
+            lock (_connectionStateLock)
+            {
+                if (_connectionState != ConnectionState.Closed)
+                {
+                    _frameWriter.WriteGoAwayAsync(_highestOpenedStreamId, Http2ErrorCode.NO_ERROR);
```

Comment:
Do you really want to do IO under the lock? Or just change the state and set a bool to send the GoAway?

Reply:
These are fire and forget so it shouldn't cause any deadlocks. Was that your concern?

Reply:
Wanting to hold the lock as briefly as possible to avoid contention or deadlocks. If they can never block then it's probably fine.